### ISSUE TITLE
build libs with `/p:Deterministic=true`  some libraray make diff

### ISSUE THREAD

**Issue by srxqds** [dotnet/runtime#76774]
I build libs with /p:Deterministic=true with the below command:
> ./build.cmd -pack -runtimeFlavor mono /p:Deterministic=true 

when I rebuild some libraries generate diff:
![image](https://user-images.githubusercontent.com/5187024/194681152-f4d21576-8add-4b06-9bd4-bf45b71b3b4b.png)

Is there tool to check they are the same code assembly?

**Comment by ViktorHofer** [comment]
@ericstj this sounds important to fix but not block .NET 7 GA. Let's investigate and schedule a fix for soon after GA to not disrupt the existing builds?

**Comment by ericstj** [comment]
The screenshot shows differences with assemblies from multiple layers of the stack.  System.Private.Xml sticks out as it is pretty low and I wouldn't expect it to rebuild.  That might be an interesting first place to look.  Essentially find the lowest component in the stack and answer the question "why did this rebuild?" keep pushing down if the answer to that question is "due to a reference changing".

**Comment by srxqds** [comment]
Is there any tool to analyze assembly diffs?

**Comment by ViktorHofer** [comment]
> Is there any tool to analyze assembly diffs?

There are many on the market, i.e. JustAssembly or assembly-differ. I usually diff them manually by looking at the ildasm output or inspecting the assemblies via ilspy.

That said, in this case I would recommend to instead look at the build logfiles which should indicate what hinders the deterministic build. Usually you would produce an msbuild bin log via the `/bl` flag and then open it with the msbuild bin log viewer.

**Comment by ViktorHofer** [comment]
> I build libs with /p:Deterministic=true with the below command:

[The deterministic property already defaults to true](https://github.com/dotnet/sdk/blob/5ea119919b69ed3cedca71600f17e940ca6d0114/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props#L34) since a few years. You don't need to pass it in yourself.

I couldn't reproduce the issue when just building the libraries:

```cmd
dotnet.cmd libs -allconfigurations && dotnet.cmd libs -allconfigurations /bl
```

The binlog of the second build doesn't show any `csc` task invocations which means the compiler wasn't invoked. I will follow your exact command later.

**Comment by ViktorHofer** [comment]
@srxqds unfortunately I can't reproduce the issue on my machine (Windows).

**Comment by srxqds** [comment]
hi, I have try again, they are always diff. I have diff with hex:
![image](https://user-images.githubusercontent.com/5187024/200764002-6210843d-c1f4-43c6-a22e-f7bd42a43181.png)

The files in zip are all diff, can you try to analyze？

[lib-diff.zip](https://github.com/dotnet/runtime/files/9968350/lib-diff.zip)

**Comment by MichalStrehovsky** [comment]
@ViktorHofer I think there's a misunderstanding - this issue is not about incremental compilation. It's about reproducible builds. I can repro the problem with latest main:

1. Delete artifacts
1. `build.cmd clr.aot+libs -rc Debug -lc Release /p:Deterministic:true`
1. Rename artifacts to artifacts.old
1. `build.cmd clr.aot+libs -rc Debug -lc Release /p:Deterministic:true`
1. `fc /b artifacts\bin\System.Private.Xml\Release\net7.0-windows\System.Private.Xml.dll artifacts.old\bin\System.Private.Xml\Release\net7.0\System.Private.Xml.dll

There should be no diffs, but fc will come up with many. Not even the file size is the same on my box (!!!).

I agree this looks like a 7.0 issue because we can't generate reproducible builds.

ILDasm on the assembly is pointing towards the Regex source generator (cc @stephentoub):

```diff
  .field static assembly initonly class System.Text.RegularExpressions.Generated.'<RegexGenerator_g>F74B1AE921BCEFE4BA601AA541C7A23B1CA9711EA81E8FE504B5B6446748E035A__Regex1_3' Instance
  .method private hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       212 (0xd4)
    .maxstack  5
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::.ctor()
    IL_0006:  ldarg.0
    IL_0007:  ldstr      "([(][(](\?<t>[^)]+)[)])\?(\?<a>[^[]+)[[](\?<ia>.+)[]]["
    + ")]\?"
    IL_000c:  stfld      string [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::pattern
    IL_0011:  ldarg.0
    IL_0012:  ldc.i4.0
    IL_0013:  stfld      valuetype [System.Text.RegularExpressions]System.Text.RegularExpressions.RegexOptions [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::roptions
    IL_0018:  ldsfld     valuetype [System.Runtime]System.TimeSpan System.Text.RegularExpressions.Generated.'<RegexGenerator_g>F74B1AE921BCEFE4BA601AA541C7A23B1CA9711EA81E8FE504B5B6446748E035A__Utilities'::s_defaultTimeout
    IL_001d:  call       void [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::ValidateMatchTimeout(valuetype [System.Runtime]System.TimeSpan)
    IL_0022:  ldarg.0
    IL_0023:  ldsfld     valuetype [System.Runtime]System.TimeSpan System.Text.RegularExpressions.Generated.'<RegexGenerator_g>F74B1AE921BCEFE4BA601AA541C7A23B1CA9711EA81E8FE504B5B6446748E035A__Utilities'::s_defaultTimeout
    IL_0028:  stfld      valuetype [System.Runtime]System.TimeSpan [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::internalMatchTimeout
    IL_002d:  ldarg.0
    IL_002e:  newobj     instance void System.Text.RegularExpressions.Generated.'<RegexGenerator_g>F74B1AE921BCEFE4BA601AA541C7A23B1CA9711EA81E8FE504B5B6446748E035A__Regex1_3'/RunnerFactory::.ctor()
    IL_0033:  stfld      class [System.Text.RegularExpressions]System.Text.RegularExpressions.RegexRunnerFactory [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::factory
    IL_0038:  ldarg.0
    IL_0039:  newobj     instance void [System.Runtime]System.Collections.Hashtable::.ctor()
    IL_003e:  dup
    IL_003f:  ldstr      "t"
    IL_0044:  ldc.i4.2
    IL_0045:  box        [System.Runtime]System.Int32
    IL_004a:  callvirt   instance void [System.Runtime]System.Collections.Hashtable::Add(object,
                                                                                         object)
    IL_004f:  dup
+    IL_0050:  ldstr      "ia"
+    IL_0055:  ldc.i4.4
-    IL_0050:  ldstr      "0"
-    IL_0055:  ldc.i4.0
    IL_0056:  box        [System.Runtime]System.Int32
    IL_005b:  callvirt   instance void [System.Runtime]System.Collections.Hashtable::Add(object,
                                                                                         object)
    IL_0060:  dup
+    IL_0061:  ldstr      "a"
+    IL_0066:  ldc.i4.3
-    IL_0061:  ldstr      "1"
-    IL_0066:  ldc.i4.1
    IL_0067:  box        [System.Runtime]System.Int32
    IL_006c:  callvirt   instance void [System.Runtime]System.Collections.Hashtable::Add(object,
                                                                                         object)
    IL_0071:  dup
+    IL_0072:  ldstr      "0"
-    IL_0072:  ldstr      "a"
+    IL_0077:  ldc.i4.0
-    IL_0077:  ldc.i4.3
    IL_0078:  box        [System.Runtime]System.Int32
    IL_007d:  callvirt   instance void [System.Runtime]System.Collections.Hashtable::Add(object,
                                                                                         object)
    IL_0082:  dup
+    IL_0083:  ldstr      "1"
-    IL_0083:  ldstr      "ia"
+    IL_0088:  ldc.i4.1
-    IL_0088:  ldc.i4.4
    IL_0089:  box        [System.Runtime]System.Int32
    IL_008e:  callvirt   instance void [System.Runtime]System.Collections.Hashtable::Add(object,
                                                                                         object)
    IL_0093:  call       instance void [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::set_CapNames(class [System.Runtime]System.Collections.IDictionary)
    IL_0098:  ldarg.0
    IL_0099:  ldc.i4.5
    IL_009a:  newarr     [System.Runtime]System.String
    IL_009f:  dup
    IL_00a0:  ldc.i4.0
    IL_00a1:  ldstr      "0"
    IL_00a6:  stelem.ref
    IL_00a7:  dup
    IL_00a8:  ldc.i4.1
    IL_00a9:  ldstr      "1"
    IL_00ae:  stelem.ref
    IL_00af:  dup
    IL_00b0:  ldc.i4.2
    IL_00b1:  ldstr      "t"
    IL_00b6:  stelem.ref
    IL_00b7:  dup
    IL_00b8:  ldc.i4.3
    IL_00b9:  ldstr      "a"
    IL_00be:  stelem.ref
    IL_00bf:  dup
    IL_00c0:  ldc.i4.4
    IL_00c1:  ldstr      "ia"
    IL_00c6:  stelem.ref
    IL_00c7:  stfld      string[] [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::capslist
    IL_00cc:  ldarg.0
    IL_00cd:  ldc.i4.5
    IL_00ce:  stfld      int32 [System.Text.RegularExpressions]System.Text.RegularExpressions.Regex::capsize
    IL_00d3:  ret
  } // end of method '<RegexGenerator_g>F74B1AE921BCEFE4BA601AA541C7A23B1CA9711EA81E8FE504B5B6446748E035A__Regex1_3'::.ctor
```

**Comment by stephentoub** [comment]
Based on that diff, it's likely due to https://github.com/dotnet/runtime/blob/2934824d2aa45ab6d558040f27de0d0027872353/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L152 and Hashtable enumeration ordering being dependent on randomized string hashcodes. I can check when I'm at my machine later today, but if that's the case, we could sort the contents as part of outpitting it.

**Comment by ViktorHofer** [comment]
Thanks @MichalStrehovsky for the clarification. If the culprit is the Regex source generator, then I would assume that any product that uses it would be affected? 

We need some form of reproducible build validation (maybe as part of the VMR's end to end tests, cc @mmitche).

**Comment by stephentoub** [comment]
Reopening to track 7.0 port

**Comment by srxqds** [comment]
thank you, found the pr merged, I will test after 7.0.1
